### PR TITLE
Fixing group assignments and other item infos which are lost in #962

### PR DESCRIPTION
**Pull request by kaikreuzer**
Signed-off-by: Kai Kreuzer kai@openhab.org

### CODE REVIEW

**Comment by sjka** [comment]
I'm working on it...

**Comment by sjka** [comment]
I created kaikreuzer#3
btw, I noticed that we should test agains the item*registry*, not the provider

**Comment by kaikreuzer** [comment]
Thanks, I included your fix in this PR, so ready to merge from my side.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ManagedItemProvider.java [hunk 1]
```diff
@@ -178,16 +130,18 @@ protected void addItemFactory(ItemFactory itemFactory) {
...
+            Iterator<String> iterator = failedToCreate.keySet().iterator();
             while (iterator.hasNext()) {
-                ItemInfo itemInfo = iterator.next();
-                Item item = itemFactory.createItem(itemInfo.itemType, itemInfo.itemName);
+                String itemName = iterator.next();
+                PersistedItem persistedItem = failedToCreate.get(itemName);
+                ActiveItem item = itemFactory.createItem(persistedItem.itemType, itemName);
                 if (item != null) {
-                    iterator.remove();
+                    failedToCreate.remove(itemName);
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ManagedItemProvider.java]: Shouldn't this break the iterator?

Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ManagedItemProvider.java]: Returns a Set view of the keys contained in this map. The set is backed by the map, so changes to the map are reflected in the set, and vice-versa. If the map is modified while an iteration over the set is in progress (except through the iterator's own remove operation), the results of the iteration are undefined.

Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ManagedItemProvider.java]: > so changes to the map are reflected in the set, and vice-versa

Is this true for a ConcurrentHashMap as well?

Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ManagedItemProvider.java]: You are right: <code>ConcurrentHashMap</code>
Returns a Set view of the keys contained in this map. The set is backed by the map, so changes to the map are reflected in the set, and vice-versa. The set supports element removal, which removes the corresponding mapping from this map, via the Iterator.remove, Set.remove, removeAll, retainAll, and clear operations. It does not support the add or addAll operations.


Shouldn't we implement against the map interface if we declarate the failedToCreate this way <code>Map&lt;String, PersistedItem&gt; failedToCreate</code>?
Sure, ConcurrentHashMap is used for the initialization, but for stability / readability in a few month I would prefer using iterator.remove that is working for the interface and the special implementation.

Comment (sjka) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ManagedItemProvider.java]: Okay, I see that point - I will fix it

**Comment by kaikreuzer** [comment]
Updated and squashed.

**Pull request merged**